db/python/tables/cohort.py

```diff
@@ -0,0 +1,198 @@
+    timestamp: GenericFilter[datetime.datetime] | None = None
+    project: GenericFilter[ProjectId] | None = None
+
+
+@dataclasses.dataclass(kw_only=True)
+class CohortTemplateFilter(GenericFilterModel):
+    """
+    Filters for CohortTemplate
+    """
+
+    id: GenericFilter[int] | None = None
+    name: GenericFilter[str] | None = None
+    description: GenericFilter[str] | None = None
+    criteria: GenericFilter[dict] | None = None
+    project: GenericFilter[ProjectId] | None = None
+
+
+class CohortTable(DbBase):
+    """
+    Capture Cohort table operations and queries
+    """
+
+    table_name = 'cohort'
+    common_get_keys = [
+        'id',
+        'name',
+        'template_id',
+        'description',
+        'author',
+        'project',
+    ]
+
+    template_keys = [
+        'id',
+        'name',
+        'description',
+        'criteria',
+        'project'
+    ]
+
+    async def query(self, filter_: CohortFilter):
+        """Query Cohorts"""
+        wheres, values = filter_.to_sql(field_overrides={})
+        if not wheres:
+            raise ValueError(f'Invalid filter: {filter_}')
+
+        common_get_keys_str = ','.join(self.common_get_keys)
+        _query = f"""
+        SELECT {common_get_keys_str}
+        FROM cohort
+        WHERE {wheres}
+        """
+
+        rows = await self.connection.fetch_all(_query, values)
+        cohorts = [Cohort.from_db(dict(row)) for row in rows]
+        return cohorts
+
+    async def get_cohort_sequencing_group_ids(self, cohort_id: int) -> list[int]:
+        """
+        Return all sequencing group IDs for the given cohort.
+        """
+
+        _query = """
+        SELECT sequencing_group_id FROM cohort_sequencing_group WHERE cohort_id = :cohort_id
+        """
+        rows = await self.connection.fetch_all(_query, {'cohort_id': cohort_id})
+        return [row['sequencing_group_id'] for row in rows]
+
+    async def query_cohort_templates(self, filter_: CohortTemplateFilter):  # TODO: Move this to its own class?
+        """Query CohortTemplates"""
+        wheres, values = filter_.to_sql(field_overrides={})
+        if not wheres:
+            raise ValueError(f'Invalid filter: {filter_}')
+
+        common_get_keys_str = ','.join(self.template_keys)
+        _query = f"""
+        SELECT {common_get_keys_str}
+        FROM cohort_template
+        WHERE {wheres}
+        """
+
+        rows = await self.connection.fetch_all(_query, values)
+        cohort_templates = [CohortTemplateModel.from_db(dict(row)) for row in rows]
+        return cohort_templates
+
+    async def get_cohort_template(self, template_id: int):
+        """
+        Get a cohort template by ID
+        """
+        _query = """
+        SELECT id as id, criteria as criteria FROM cohort_template WHERE id = :template_id
+        """
+        template = await self.connection.fetch_one(_query, {'template_id': template_id})
+
+        if not template:
+            return None
```

Comment:
Maybe :shrug: We went backwards and forwards on this — see this thread: https://github.com/populationgenomics/metamist/pull/615/files#r1542111253

I was thinking if writing a query at table level, I would want to be informed if the `template_id` I was asking about didn't exist rather than have to handle that as an error. However looking at the nearby code, e.g. `get_analysis_by_id`, it does look like the convention here is that these APIs expect to be given known-to-be-valid IDs so raising NotFoundError is probably better.

Reply:
Yeah tough, I don't mind this convention, as long as it's typed or clear. But it is semantically different to our other functions.

Reply:
Have switched it to a NotFoundError for now. Give me a shout if anyone feels strongly about moving back to the return None implementation.